infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java

```diff
@@ -46,58 +56,140 @@
+   * no-placeholders |       no            |   no-placeholders    | no
+   * no-placeholders |       no            |    placeholders      | yes
+   * no-placeholders |      yes            |       null           | no
+   * no-placeholders |      yes            |   no-placeholders    | no
+   * no-placeholders |      yes            |    placeholders      | no
+   *  placeholders   |       no            |        null          | error
+   *  placeholders   |       no            |   no-placeholders    | no
+   *  placeholders   |       no            |    placeholders      | yes
+   *  placeholders   |      yes            |        null          | yes
+   *  placeholders   |      yes            |   no-placeholders    | yes
+   *  placeholders   |      yes            |    placeholders      | yes
+   * }</pre>
    */
-  public boolean isPredefined() {
-    return isPredefined;
+  protected boolean isCreatingNamespace(String workspaceId) throws InfrastructureException {
+    boolean legacyExists =
+        checkNamespaceExists(
+            resolveLegacyNamespaceName(EnvironmentContext.getCurrent().getSubject(), workspaceId));
+
+    // legacy namespace exists and should be used
+    if (legacyExists) {
+      // if it contains any placeholder("" is <workspaceid>) - it indicates that Che created
+      // namespace by itself
+      return isNullOrEmpty(legacyNamespaceName) || hasPlaceholders(legacyNamespaceName);
+    }
+
+    if (isNullOrEmpty(defaultNamespaceName)) {
+      throw new InfrastructureException(
+          "Cannot determine whether a new namespace and service account should be"
+              + " created for workspace %s. There is no pre-existing workspace namespace to be"
+              + " found using the legacy `che.infra.kubernetes.namespace` property yet the"
+              + " `che.infra.kubernetes.namespace.default` property is undefined.");
+    }
+
+    return hasPlaceholders(defaultNamespaceName);
+  }
+
+  /**
+   * Returns true if the namespace is fully managed by Che (e.g. created, used and deleted), false
+   * otherwise.
+   */
+  public boolean isManagingNamespace(String workspaceId) throws InfrastructureException {
+    // the new logic is quite simple.
+    boolean newLogic =
+        defaultNamespaceName != null && defaultNamespaceName.contains(WORKSPACEID_PLACEHOLDER);
+
+    // but we must follow the same logic as #evalNamespaceName - we need to make sure that the old
+    // logic can't be used first...
+    // empty legacy namespace name ~ <workspaceid>
+    if (isNullOrEmpty(legacyNamespaceName)
+        || legacyNamespaceName.contains(WORKSPACEID_PLACEHOLDER)) {
+
+      // there's a chance of using the old logic - if the namespace exists, we're managing it.
+      // if it doesn't, we're using the new logic.
+      return checkNamespaceExists(
+              resolveLegacyNamespaceName(EnvironmentContext.getCurrent().getSubject(), workspaceId))
+          || newLogic;
```

Comment:
we use || and it means that is newLogic is true - then it does not matter if namespace exists or not.
It makes me thinking that something is wrong. Please double check that it's correct

Reply:
Seems we can override it in the following way:
```java
  public boolean isManagingNamespace(String workspaceId) throws InfrastructureException {
    // the new logic is quite simple.
    boolean isNewNamespaceManaged =
        defaultNamespaceName != null && defaultNamespaceName.contains(WORKSPACEID_PLACEHOLDER);
    
    if (isNewNamespaceManaged) {
      return true;
    }

    boolean legacyHasWorkspacePlaceholder = isNullOrEmpty(legacyNamespaceName) || legacyNamespaceName.contains(WORKSPACEID_PLACEHOLDER);
    return legacyHasWorkspacePlaceholder && checkNamespaceExists(resolveLegacyNamespaceName(EnvironmentContext.getCurrent().getSubject(), workspaceId));
  }
```
Not sure if it's simpler for understanding but if conditions seems clearer.